testnet/stacks-node/src/tests/epoch_205.rs

```diff
@@ -913,3 +916,287 @@ fn test_cost_limit_switch_version205() {
+    wait_for_microblocks(&microblocks_processed, 180);
+
+    microblocks_processed.store(0, Ordering::SeqCst);
+
+    // this test can sometimes miss a mine block event.
+    sleep_ms(80_000);
+    next_block_and_wait(&mut btc_regtest_controller, &blocks_processed);
+
+    let blocks = test_observer::get_blocks();
+    assert!(blocks.len() >= 5, "Should have produced at least 5 blocks");
+
+    // max == largest number of transactions per stream in a given epoch (2.0 or 2.05)
+    // total == number of transactions across all streams in a given epoch (2.0 or 2.05)
+    let mut max_big_txs_per_microblock_20 = 0;
+    let mut total_big_txs_per_microblock_20 = 0;
+
+    let mut max_big_txs_per_microblock_205 = 0;
+    let mut total_big_txs_per_microblock_205 = 0;
+
+    let mut in_205 = false;
+
+    // NOTE: this only counts the number of txs per stream, not in each microblock
+    for block in blocks {
+        let transactions = block.get("transactions").unwrap().as_array().unwrap();
+        eprintln!("{}", transactions.len());
+
+        let mut num_big_microblock_txs = 0;
+
+        for tx in transactions.iter() {
+            let raw_tx = tx.get("raw_tx").unwrap().as_str().unwrap();
+            if raw_tx == "0x00" {
+                continue;
+            }
+            let tx_bytes = hex_bytes(&raw_tx[2..]).unwrap();
+            let parsed = StacksTransaction::consensus_deserialize(&mut &tx_bytes[..]).unwrap();
+            if let TransactionPayload::SmartContract(tsc) = parsed.payload {
+                if tsc.name.to_string().find("costs-2").is_some() {
+                    in_205 = true;
+                } else if tsc.name.to_string().find("large").is_some() {
+                    num_big_microblock_txs += 1;
+                    if in_205 {
+                        total_big_txs_per_microblock_205 += 1;
+                    } else {
+                        total_big_txs_per_microblock_20 += 1;
+                    }
+                }
+            }
+        }
+        if in_205 && num_big_microblock_txs > max_big_txs_per_microblock_205 {
+            max_big_txs_per_microblock_205 = num_big_microblock_txs;
+        }
+        if !in_205 && num_big_microblock_txs > max_big_txs_per_microblock_20 {
+            max_big_txs_per_microblock_20 = num_big_microblock_txs;
+        }
+    }
+
+    eprintln!(
+        "max_big_txs_per_microblock_20: {}, total_big_txs_per_microblock_20: {}",
+        max_big_txs_per_microblock_20, total_big_txs_per_microblock_20
+    );
+    eprintln!(
+        "max_big_txs_per_microblock_205: {}, total_big_txs_per_microblock_205: {}",
+        max_big_txs_per_microblock_205, total_big_txs_per_microblock_205
+    );
+
+    assert!(max_big_txs_per_microblock_20 <= 1);
+    assert!(total_big_txs_per_microblock_20 <= 1);
+
+    // can be 9; allow a fudge factor
+    assert!(max_big_txs_per_microblock_205 >= 8);
+    assert!(total_big_txs_per_microblock_205 >= 8);
```

Comment:
Is testing the transaction receipt costs instead of testing the tx count sufficient? That is, assert that the big transactions in the microblock consume X cost before and Y cost after the epoch switch? That's a little more robust of an assertion which should cut down on test flapping and sensitivity to the runtime environment, but if its important to capture the block limit itself in the assertion somehow, perhaps just test that the sum of the MARF operations is less than the new limit but greater than the old limit?

Reply:
Sure, done.